doc/dev/architecture.rst

```diff
@@ -5,71 +5,392 @@ Architecture
+  `primitives <https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html>`_).
+  Custom JAXPR -> MLIR lowerings are registered to these primitives to fully convert a hybrid
+  program to MLIR for consumption by the compiler.
 
-|br|
+- **Program Transformation:** The main part of compilation is performed on the MLIR-based
+  representation for hybrid quantum programs defined by Catalyst. The user program is passed to
+  the Catalyst compiler libraries in its textual form, as the MLIR memory objects are not
+  compatible between Catalyst and ``jaxlib``. The driver then invokes a sequence of transformations
+  that *lowers* the user program to a lower level of abstraction, outputting LLVM IR with QIR
+  syntax.
+  For more details consult the :ref:`next section<compiler>`.
+
+- **Code Generation:** At this stage the LLVM IR is compiled down to native object code using the
+  LLVM Static Compiler (``llc``) for the local system architecture. A native linker is then used
+  to link the user program to the Catalyst Runtime library. The frontend will load this library
+  into the Python environment and attach its entry point to the callable ``@qjit`` object defined
+  by the user.
+  For more details on the program execution consult the :ref:`runtime section<runtime>`.
+
+Elaborating on the program capture phase, tracing is a mechanism by which a function is executed
+with abstract arguments called *tracers*. Calling operations from the tracing library (such as
+``jax.numpy``) will record the operation and which tracers (or constants) it acts upon in a global
+data structure. Note that JAX's tracing contexts are thread-local and can be nested to allow scoped
+region capture, which is relevant when tracing control flow operations.
+
+During the tracing of quantum functions (``qml.qnode``), PennyLane's queuing context is activated to
+build a ``QuantumTape`` data structure that records all quantum operations. Nested queuing contexts
+are leveraged to allow for scoped operation capture, including control flow operations which are
+themselves captured as pseudo-quantum operations on the tape.
+
+Catalyst provides the "glue" to embed quantum tapes into the JAXPR, by converting PennyLane
+operations to their corresponding JAX primitive and by connecting operation arguments/results to
+the correct tracer objects.
 
 
+.. _compiler:
+
+Compiler Core
+^^^^^^^^^^^^^
+
+This section will focus on the series of compiler passes that convert a high-level quantum program
+to its low-level LLVM IR and subsequent binary form. While the exact passes may frequently change,
+the general stages should still be applicable.
+See the graph below for an overview of the transformations applied to the user program:
+
+|br|
+
 .. image:: ../_static/arch/compiler.svg
-  :width: 650
-  :alt: Compiler Workflow
+  :width: 800
+  :alt: Compiler Core
   :align: center
 
 |br|
 
-.. _Legend:
+- **HLO lowering:**
```

Comment:
@dime10 since these are numbered steps, should they be instead?
```suggestion
1. **HLO lowering:**
```

(and same for the ones that follow)

Reply:
Note if this change is accepted, you will need to indent by an extra space the nested bullet points below :(

Reply:
Well they mostly are ordered but not necessarily (e.g. the way quantum and classical optimizations appear is not necessarily implying a fixed order). I would prefer to keep an unordered list :)